![Johnson & Johnson MedTech Logo](https://raw.githubusercontent.com/Forage-Simulations/Johnson-Johnson-Robotics-Controls/main/github_assets.png)


# Control System Diagnostic Notebook for Robotic Arm

### Objective
This notebook is designed to help identify and resolve response delays in the control code of a robotic arm. You will:

- Diagnose the root cause of delays in command response times.
- Optimize control code for improved performance.
- Document your findings and propose actionable solutions.

### Instructions
Follow the steps outlined in this notebook to diagnose and resolve issues in the robotic arm's control system:

1. **Load Libraries**: Run the provided setup code to load necessary Python libraries.
2. **Run Diagnostic Functions**: Use the `check_response_time` function to measure command response times.
3. **Analyze Findings**: Record observed delays and propose a hypothesis for their cause.
4. **Test Optimizations**: Apply optimization logic and compare results.
5. **Record Results**: Summarize your findings and recommendations in a structured format.

Focus on the `rotate_joint` command, as it has been flagged for delays.


In [1]:
# Import required libraries
import time  # For measuring response times
import numpy as np  # For numerical calculations


In [2]:
# Define a function to measure the response time of commands
def check_response_time(command):
    """Simulates command execution and measures response time."""
    start_time = time.time()
    if command == "rotate_joint":
        time.sleep(0.15)  # Simulate delay for rotate_joint
    elif command == "move_arm":
        time.sleep(0.1)  # Simulate moderate response time
    elif command == "adjust_grip":
        time.sleep(0.05)  # Simulate fast response time
    response_time = time.time() - start_time
    return response_time


In [3]:
# List of commands to test
commands = ["move_arm", "rotate_joint", "adjust_grip"]

# Measure and print response times for each command
print("Testing initial command response times:")
for cmd in commands:
    response_time = check_response_time(cmd)
    print(f"{cmd} response time: {round(response_time, 3)} seconds")


Testing initial command response times:
move_arm response time: 0.1 seconds


rotate_joint response time: 0.15 seconds
adjust_grip response time: 0.05 seconds


### Step 3: Analyze Initial Findings

Record the response times observed for each command. Focus on identifying commands with higher response times.

| Command       | Observed Response Time | Expected Response Time | Notes on Performance   |
|---------------|------------------------|------------------------|-------------------------|
| move_arm      |        0.1             | 0.10                   |  MATCHES AS EXPECTED    |
| rotate_joint  |        0.15            | 0.18                   |  MEASURED IS FASTER 
|               |                         |                         |  THAN THE REPORTED DELAY|    
| adjust_grip   |        0.05            | 0.09                   |FASTER THAN EXPECTED     |

#### Hypothesis:
The `rotate_joint` delay is likely caused by more complex positional calculation than the other commands, and by re-executing its full computation on every call even when the requested joint angle is unchanged, so part of the delay is redundant work. It is the slowest command (0.15 s) and the one flagged in ticket #2437. The other two commands behave as expected.

*Note:* measured `rotate_joint` (0.15 s) is faster than the 0.18 s reported in the ticket. The reported figure is treated as the expected/threshold value, so run-to-run variance or load on the reporting system may explain the gap.


In [4]:
# Define a function to simulate optimized command execution
def optimized_command(command, improvement_factor=0.2):
    """Simulates optimized command execution."""
    print(f"Optimizing command: {command}")  # Placeholder action
    optimized_response_time = check_response_time(command) * (1 - improvement_factor)
    return optimized_response_time


In [5]:
# Test each command after optimizations
print("\nTesting optimized command response times:")
for cmd in commands:
    optimized_time = optimized_command(cmd)
    print(f"{cmd} optimized response time: {round(optimized_time, 3)} seconds")



Testing optimized command response times:
Optimizing command: move_arm
move_arm optimized response time: 0.08 seconds
Optimizing command: rotate_joint


rotate_joint optimized response time: 0.12 seconds
Optimizing command: adjust_grip
adjust_grip optimized response time: 0.04 seconds


### Step 4b: Mechanism-based optimization (`JointCommander`)

The flat 20% factor above is only an assumption. Here the optimization is implemented for real: `JointCommander` remembers the last joint target, and if `rotate_joint` is called again with the same target (within a small tolerance) it skips the recomputation. New targets still run at full cost.

The workload is 10 joint commands in which 2 repeat the previous target, which is typical when a controller re-issues a position it is already holding.


In [6]:
class JointCommander:
    """Caches the last joint target so repeated identical targets skip recomputation."""
    def __init__(self, tolerance=1e-3):
        self.tolerance = tolerance
        self.last_target = None

    def rotate_joint(self, target):
        start = time.time()
        if self.last_target is not None and abs(target - self.last_target) <= self.tolerance:
            pass  # cache hit: joint is already at this target, nothing to recompute
        else:
            check_response_time("rotate_joint")  # full-cost rotation for a new target
            self.last_target = target
        return time.time() - start


targets = [10, 10, 25, 40, 40, 55, 70, 85, 100, 120]  # degrees; 2 of 10 repeat the previous target

baseline = [check_response_time("rotate_joint") for _ in targets]
commander = JointCommander()
cached = [commander.rotate_joint(t) for t in targets]

print("target  baseline  cached")
for t, b, c in zip(targets, baseline, cached):
    print(f"{t:>6}  {b:8.3f}  {c:6.3f}")

avg_base, avg_cached = np.mean(baseline), np.mean(cached)
print(f"\nAverage rotate_joint response time: {avg_base:.3f} s -> {avg_cached:.3f} s "
      f"({(1 - avg_cached / avg_base) * 100:.0f}% faster)")

# Genuinely new targets must not be affected by the cache
new_only = [c for t, c, prev in zip(targets[1:], cached[1:], targets[:-1]) if t != prev]
print(f"Average for new targets only: {np.mean(new_only):.3f} s (unchanged from baseline)")

target  baseline  cached
    10     0.150   0.150
    10     0.150   0.000
    25     0.150   0.150
    40     0.150   0.150
    40     0.150   0.000
    55     0.150   0.150
    70     0.150   0.150
    85     0.150   0.150
   100     0.150   0.150
   120     0.150   0.150

Average rotate_joint response time: 0.150 s -> 0.120 s (20% faster)
Average for new targets only: 0.150 s (unchanged from baseline)


### Step 5: Record Results

#### Observations:
| Command | Initial (s) | Optimized (s) | Change |
|---|---|---|---|
| move_arm | 0.10 | 0.08 | -20% |
| rotate_joint | 0.15 | 0.12 | -20% |
| adjust_grip | 0.05 | 0.04 | -20% |

#### Key Insights:
- The table above uses the flat 20% factor from `optimized_command`, which is an assumption.
- The real mechanism (`JointCommander`, Step 4b) was measured on a 10-command workload with 2 repeated targets: average `rotate_joint` time went from 0.15 s to about 0.12 s (see cell output). Repeats cost about 0 s, new targets still cost the full 0.15 s.
- `rotate_joint` remains the slowest command after optimization (0.12 s vs 0.08 s and 0.04 s), so it is still the best target for real profiling.
- The absolute gain is largest for `rotate_joint` (0.03 s per call), which matters most for repeated joint moves.


### Step 6: Summary and Recommendations

**Identified Issue:** `rotate_joint` showed the longest response time (0.15 s), consistent with the delay flagged in the ticket. The likely cause is redundant calculations and inefficient loops in its control code.

**Optimization Applied:** Implemented a caching mechanism (`JointCommander`) that stores the last-issued joint target and skips recomputation when a call repeats the same target. Genuinely new targets still run at full cost. Measured on a 10-command workload with 2 repeated targets: average `rotate_joint` time 0.15 s to about 0.12 s.

**Next Steps:**
- Test `JointCommander` on real command sequences and hardware, since the workload here is simulated.
- Validate the caching logic against real hardware feedback to ensure joint position hasn't drifted.
- Profile the underlying kinematic computation (e.g. `cProfile`) if delays persist for genuinely new targets.
- Replace the simulated 20% factor with measured before/after timings, averaged over many runs.
- Stress-test under high load to check stability.
- Add automated response-time checks with a threshold so regressions are flagged early.


---
# Diagnostic Report: Robotic Arm Control System

**Title:** Diagnostic analysis and optimization of robotic arm control system
**Name:** Pavan Kushal Velagaleti  
**Date:** 2026-09-19  
**Ticket ID:** #2437

## Diagnostic process
**Initial observations:** `rotate_joint` was the slowest command (0.15 s), matching the delay described in the ticket.

**Commands tested:** `move_arm`, `rotate_joint`, `adjust_grip`

**Hypothesis:** `rotate_joint` re-executes its full computation on every call, even when the requested joint angle is unchanged, so part of the delay is redundant.

**Tools and techniques:** Python, this diagnostic notebook, a response-time measurement function (`check_response_time`), iterative testing.

## Findings and analysis
| Command | Expected (s) | Initial (s) | Optimized (s) |
|---|---|---|---|
| move_arm | 0.10 | 0.10 | 0.08 |
| rotate_joint | 0.18 (ticket) | 0.15 | 0.12 |
| adjust_grip | 0.09 | 0.05 | 0.04 |

`rotate_joint` is the slowest command and the most likely place for redundant logic. `move_arm` matched expectations and `adjust_grip` was faster than expected. Delays here are simulated with `time.sleep`, so the analysis identifies where to look, not a confirmed root cause.

## Optimizations and solutions
- **`rotate_joint`:** caching (`JointCommander`) of the last joint target to avoid redundant recomputation on repeated identical targets.
- **Impact:** on a 10-command workload with 2 repeated targets, average `rotate_joint` time fell from 0.15 s to about 0.12 s (measured in Step 4b); new targets still cost the full 0.15 s. The Step 4 table for `move_arm` (0.10 to 0.08) and `adjust_grip` (0.05 to 0.04) uses an assumed flat 20% factor, not a measured change.

## Recommendations
- **Preventive:** periodic audits of control code; automated response-time tests that flag regressions.
- **Further testing:** validate the cache against real hardware feedback (joint drift); repeated trials under realistic load; profile the kinematic computation if delays persist for new targets.

## Conclusion
Diagnostics identified `rotate_joint` as the bottleneck (0.15 s), and the `JointCommander` cache brings the average to about 0.12 s on a workload with repeated targets. **Next steps:** validate on real hardware and realistic load, deploy, then monitor performance in production.
